example/libp2p-chat-with-circuit-relay.ss

```diff
@@ -0,0 +1,169 @@
+        (displayln "*** STREAM CLOSED"))
+       ((string-empty? line)
+        (lp))
+       (else
+        (write-u8 #x1b)
+        (display "[32m")
+        (display line)
+        (write-u8 #x1b)
+        (displayln "[0m")
+        (display "> ")
+        (lp))))))
+
+;; executable program
+(def (main . args)
+  (def listen-cmd
+    (command 'listen help: "listen for incoming connections"
+             (argument 'host-addresses help: "Comma separated multi addresses")
+             (argument 'circuit-relay value: string->peer-info help: "Multi address of the circuit relay")))
+  (def dial-cmd
+    (command 'dial help: "dial an echo peer"
+             (argument 'peer value: string->peer-info help: "Address of peer to dial")
+             (argument 'host-addresses  help: "Comma separated multi addresses")))
+  (def gen-circuit-relay-cmd
+    (command 'gen-circuit-relay help: "generate a circuit relay to connect to"
+             (argument 'host-addresses help: "Comma seperated multiaddresses to host at")))
+  (def help-cmd
+    (command 'help help: "display help; help <command> for command help"
+             (optional-argument 'command value: string->symbol)))
+  (def gopt
+    (getopt listen-cmd dial-cmd gen-circuit-relay-cmd))
+
+  (start-logger!)
+  (try
+   (let ((values cmd opt) (getopt-parse gopt args))
+     (let-hash opt
+       (case cmd
+         ((listen)
+          (do-listen .host-addresses .circuit-relay))
+         ((dial)
+          (do-dial .peer .host-addresses))
+         ((gen-circuit-relay)
+          (do-circuit-relay .host-addresses))
+         ((help)
+          (getopt-display-help-topic gopt .?command "libp2p-chat")))))
+   (catch (getopt-error? exn)
+     (getopt-display-help exn "libp2p-chat" (current-error-port))
+     (exit 1))))
+
+(def (do-listen host-addresses circuit-relay)
+  (let* ((c (open-libp2p-client host-addresses: host-addresses wait: 20))
+         (self (libp2p-identify c)))
+    (for (p (peer-info->string* self))
+      (displayln "I am " p))
+    (displayln "Listening for incoming connections")
+    (libp2p-listen c [chat-proto] chat-handler)
+    (displayln "Connecting to Circuit Relay")
+    (libp2p-connect c circuit-relay) ;;This allows for the dialing peer to dial the listener through the circuit relay
+    (displayln "Connected to Circuit Relay")
+    (thread-sleep! +inf.0)))
+
+(def (do-dial peer host-addresses)
+  (displayln "dialing")
+  (let* ((c (open-libp2p-client host-addresses: host-addresses wait: 20))
+         (self (libp2p-identify c)))
+    (for (p (peer-info->string* self))
+      (displayln "I am " p))
+    (displayln "Connecting to " (peer-info->string peer))
+    (libp2p-connect c peer)
+    (let (s (libp2p-stream c peer [chat-proto]))
+      (do-chat s))))
+
+;;start a libp2p daemon that allows for circuit relays
+(def (do-circuit-relay host-addresses)
+  (let* ((d (start-libp2p-daemon! host-addresses: host-addresses options: ["-relayActive" "-relayHop"] wait: 20))
```

Comment:
you dont need the relayActive flag, thats for active relay dialing.

Reply:
Sounds good! I'll fix that. Is there a place I can look for a good description of these flags? I've been reading the source code for the daemon and it's not the most helpful

Reply:
not really :(